swarm/CHANGELOG.md

```diff
@@ -1,3 +1,16 @@
+# 0.40.0 [unreleased]
+
+- Use the total number of alive inbound streams for back-pressure. This can have a BIG impact on your application
+  depending on how it uses `libp2p`. Previously, the limit for inbound streams per connection only applied to the
+  _upgrade_ phase, i.e. for the time `InboundUpgrade` was running. Any stream being returned from `InboundUpgrade` and
+  given to the `ConnectionHandler` did not count towards that limit, essentially mitigating the back-pressure mechanism.
+  With this release, substreams count towards that limit until they are dropped and thus we actually enforce, how many
+  inbound streams can be active at one time _per connection_. `libp2p` will not accept any more incoming streams once
+  that limit is hit. If you experience stalls or unaccepted streams in your application, consider upping the limit via
+  `SwarmBuilder::max_negotiating_inbound_streams`. See [PR 2878].
```

Comment:
I have to give this more thought. Previously I was under the impression that this will be  feature for reporting only, but not actually enforcing a limit.

As of today I am undecided whether the number of established inbound streams should have a global maximum, or whether it should only have a maximum per `ConnectionHandler` implementation, potentially coordinated with the `NetworkBehaviour` and thus a maximum per `NetworkBehaviour` implementation. I am not sure the global application has enough knowledge to set a suitable limit for all protocols, nor do I think a static limit is a good idea in the first place.

Ideally I would like to enforce a limit for connections only. With https://github.com/libp2p/rust-libp2p/pull/2828 a connection limit could e.g. be dynamic based on the current memory utilization. Protocols, e.g. like Kademlia, would enforce a streams-per-connection limit in their `ConnectionHandler` implementation. That limit would be a value of maximum expected parallelization, e.g. 16 (as in "we don't expect an implementation to be handle more than 16 requests in parallel").

What do you think @thomaseizinger?

Reply:
If we want the limit to be dynamic, all we need to do is add a function to `ConnectionHandler` that allows us to query the max number of allowed streams:

```rust
trait ConnectionHandler {
	fn max_inbound_streams(&self) -> usize {
		128 // This is today's default.
	}
}
```

We can then use this limit on every iteration of `Connection::poll` to check if we should poll for more inbound streams.

If we move forward with https://github.com/libp2p/rust-libp2p/issues/2863 then the notion of upgrades will be gone and this limit does not mean anything anymore. Thus, I am inclined to say we should remove this from the `Swarm` and `Pool` altogether and only have the `ConnectionHandler` decide.

Reply:
> If we move forward with #2863 then the notion of upgrades will be gone and this limit does not mean anything anymore. Thus, I am inclined to say we should remove this from the `Swarm` and `Pool` altogether and only have the `ConnectionHandler` decide.

That would be my preference.

Reply:
LMK what you think of [`40ab076` (#2878)](https://github.com/libp2p/rust-libp2p/pull/2878/commits/40ab076ac88a16ec7e9c647935725b92a764753a).